### DISCUSSION TITLE
can we develop login page without using FTL and react?

### DISCUSSION THREAD

**Discussion by dtbdtb** [keycloak/keycloak#33158] [Q&A]
hi,evryone:
we are not familar with FTL and react .
is keycloak must open page use FTL framework?
we find keycloak internal open page in FTL framework.
so could anyone kindly tell us can we develop login page without using FTL and react?
thank you !

**Comment by jonkoops** [comment]
You will have to use FreeMarker templates to some degree as they are used to bootstrap the HTML even when React is on the page. I would suggest you learn these technologies if you want to customize Keycloak, or you will be doing a lot more work if you want to build things from scratch. That said, there is nothing preventing you from using another stack for the front-end than React. If you are looking for an easier way to customize Keycloak perhaps [Keycloakify](https://www.keycloakify.dev/) is a good option.

**Comment by ssilvert** [comment]
To add to what Jon said, there is nothing that prevents you from using whatever technology you like to create login pages or any other of our UIs such as admin and account consoles.

The only real requirement is that you use a file called `index.ftl` as the bootstrap for consoles, `login.ftl` for login, etc.  You can code the FTL as an ordinary html page if you want.

But Jon is right that it will make your life a lot easier if you at least familiarize yourself with FTL.  It's a very simple templating language.
